File: CodingInterview/Programmers/Python/LV1/42_음양_더하기.py
# https://school.programmers.co.kr/learn/courses/30/lessons/76501?language=python3

# Number of times solving questions: 0
# 2026-00-00

import sys, pathlib
sys.path.append(str(pathlib.Path(__file__).resolve().parent.parent))  # test_runner 경로
from test_runner import run_tests

# ─────────────────────────────────────────────────────────────
# 문제 규칙
# ─────────────────────────────────────────────────────────────
#   정수의 절댓값 배열 absolutes와 부호 배열 signs가 주어진다.
#   - signs[i]가 True면 absolutes[i]는 양수, False면 음수다.
#   - 모든 수를 실제 부호로 복원해 더한 값을 반환한다.
#   - 예) absolutes=[4,7,12], signs=[T,F,T] → 4 - 7 + 12 = 9

# ─────────────────────────────────────────────────────────────
# 풀이: 부호에 따라 ± 누적
# ─────────────────────────────────────────────────────────────
# 핵심 아이디어:
#   인덱스를 돌며 signs[i]가 참이면 더하고, 거짓이면 뺀다.
#   (불리언은 `== True` 비교 없이 `if signs[i]:`로 바로 판정한다.)
def solution(absolutes, signs):
    result = 0
    len_signs = len(signs)

    for i in range(len_signs):
        if signs[i]:               # 양수면 더하고
            result += absolutes[i]
        else:                      # 음수면 뺀다
            result -= absolutes[i]

    return result

# 더 파이썬다운 대안 (인덱스 없이 두 배열을 짝지어 합산):
#   def solution(absolutes, signs):
#       return sum(a if s else -a for a, s in zip(absolutes, signs))
#   → zip으로 (절댓값, 부호) 쌍을 동시 순회 + 삼항으로 부호 적용 + sum 한 번에.

# ─────────────────────────────────────────────────────────────
# 복잡도 분석
# ─────────────────────────────────────────────────────────────
# 시간: O(n) — 모든 원소를 한 번씩 더하는 게 하한
# 공간: O(1) — 누적 변수만 사용
# 결론: 이미 최적 복잡도. 차이는 성능이 아니라 "표현"의 영역.

# ─────────────────────────────────────────────────────────────
# 이 문제에서 배운 것 (Takeaways)
# ─────────────────────────────────────────────────────────────
#   1. 두 시퀀스를 동시에 돌 땐 range(len())+인덱싱보다 zip이 정석이다.
#   2. 불리언은 `if x == True`가 아니라 `if x:`로 바로 판정한다(안티패턴 회피).
#   3. "부호 적용 후 합산"은 sum(a if s else -a for ...)로 한 줄에 표현 가능.

# ─────────────────────────────────────────────────────────────
# 테스트
# ─────────────────────────────────────────────────────────────
test_cases = [
    (([4, 7, 12], [True, False, True]), 9),
    (([1, 2, 3], [False, False, True]), 0),
]

run_tests(solution, test_cases)

File: CodingInterview/Programmers/Cpp/LV1/42_음양_더하기.cpp
// https://school.programmers.co.kr/learn/courses/30/lessons/76501?language=cpp
// 출처: 월간 코드 챌린지 시즌2
// 대응 Python 풀이: ../../Python/LV1/42_음양_더하기.py

// Number of times solving questions: 1
// 2026-09-10

#include "../test_runner.h"
// 아래는 프로그래머스 제출 코드와 동일하게 유지한다.
#include <string>
#include <vector>
using namespace std;

int solution(vector<int> absolutes, vector<bool> signs) {
    // 규칙 1. absolutes[i] 의 부호는 signs[i] 가 정한다. true 면 양수, false 면 음수.
    // 규칙 2. 실제 정수들의 합을 반환한다.
    // 규칙 3. 길이·원소 모두 1000 이하 → 합의 최댓값이 1000 * 1000 = 100만.
    //         int(약 21억)로 충분하므로 long long 이 필요 없다. (59번과 반대 판단)

    int answer = 0;

    // size() 는 unsigned 라 int 와 직접 비교하면 -Wsign-compare 경고가 난다.
    int signs_size = signs.size();

    // 두 배열을 같은 인덱스로 짝지어 봐야 하므로 range-for 가 아니라 인덱스 루프를 쓴다.
    //
    // vector<bool> 은 다른 vector 와 달리 비트 단위로 압축 저장되는 특수화 타입이다.
    // 여기서는 읽기만 하므로 문제없지만, bool& 로 참조를 받으려 하면 컴파일되지 않는다.
    for (int i = 0; i < signs_size; i++) {
        if (signs[i]) {
            answer += absolutes[i];
        } else {
            answer -= absolutes[i];     // (규칙 1)
        }
    }

    return answer;
}

int main() {
    Tester t("solution");
    t.check(solution({4, 7, 12}, {true, false, true}), 9);
    t.check(solution({1, 2, 3}, {false, false, true}), 0);

    // 아래는 직접 추가한 경계 케이스
    t.check("원소 하나", solution({5}, {false}), -5);
    t.check("전부 음수", solution({1, 2, 3}, {false, false, false}), -6);
    t.check("전부 양수", solution({1, 2, 3}, {true, true, true}), 6);
    return t.report();
}
